pkg/log/log.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+
+	"github.com/vmware/vic/pkg/log/syslog"
+)
+
+type LoggingConfig struct {
+	Formatter logrus.Formatter
+	Level     logrus.Level
+	Syslog    *syslog.SyslogConfig
+}
+
+var initializer struct {
+	once sync.Once
+	err  error
+}
+
+func NewLoggingConfig() *LoggingConfig {
+	return &LoggingConfig{
+		Formatter: NewTextFormatter(),
+		Level:     logrus.InfoLevel,
+	}
+}
+
+func Init(cfg *LoggingConfig) error {
+	initializer.once.Do(
+		func() {
+			var err error
+
+			logger := logrus.StandardLogger()
+			f := logger.Formatter
+			l := logger.Level
+			defer func() {
+				initializer.err = err
+				if err != nil {
+					// revert
+					logrus.SetFormatter(f)
+					logrus.SetLevel(l)
+				}
+			}()
+
+			logrus.SetFormatter(cfg.Formatter)
+			logrus.SetLevel(cfg.Level)
+
+			if cfg.Syslog != nil {
+				cfg.Syslog.Tag = syslog.MakeTag(cfg.Syslog.Tag)
```

Comment:
I did play with that a little bit, but then we have to decide what to truncate -- the process name or the vch name, since tag is limited to 32 characters by syslog. I went with just the proc name since syslog messages do have hostnames in them, and that can be used to distinguish VCH's. However, if the ip address changes then of course this becomes a little bit untenable.